### PR TITLE
Add a sample Tensorflow function

### PR DESCRIPTION
**Pull request by chzhyang**
# Changes

This sample HTTP function is used for real-time inference(image recognition) with TensorFlow on a pretrained model(ResNet50).

- The function(`func.py`) is based on class `ImageRecognitionService`(`image_recognition_service.py`), and exposes GET, POST interface for inference.
- Class `ImageRecognitionService` provides several functions for data preprocessing, running inference and inference result parsing using TensorFlow.

### CODE REVIEW

File: tensorflow/image-recognition/func.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+specVersion: 0.34.0
```
Comment (matejvasek) [inline tensorflow/image-recognition/func.yaml:1]: The `func.yaml` config shouldn't be part of "template" AFAIK.
If any values need default value `manifest.yaml`  should be used.

File: tensorflow/image-recognition/func.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import os
+from http.client import BAD_REQUEST
```
Comment (matejvasek) [inline tensorflow/image-recognition/func.py:2]: Is this different from `werkzeug.exceptions.BadRequest` which is also imported?

File: tensorflow/image-recognition/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+parliament-functions==0.1.0
+--index-url=https://pypi.python.org/simple
+intel-tensorflow==2.9.1
```
Comment (matejvasek) [inline tensorflow/image-recognition/requirements.txt:3]: I know absolutely nothing about tensorflow.
Is there difference between `intel-tensorflow` and `tensorflow` packages?
Would `intel-tensorflow` work on ARM machine?

File: tensorflow/image-recognition/func.yaml (hunk 1 above)
Comment (chzhyang) [reply tensorflow/image-recognition/func.yaml:1]: I will update it to default value.

File: tensorflow/image-recognition/func.py (hunk 2 above)
Comment (chzhyang) [reply tensorflow/image-recognition/func.py:2]: Thanks, they are the same, I will delete one :)

File: tensorflow/image-recognition/requirements.txt (hunk 3 above)
Comment (chzhyang) [reply tensorflow/image-recognition/requirements.txt:3]: The [intel-tensorflow](https://pypi.org/project/intel-tensorflow/2.9.1/) was optimized on [tensorflow](https://pypi.org/project/tensorflow/2.9.1/),  it supports all x86 platforms and the performance is better, but it does not support ARM.

File: tensorflow/image-recognition/func.yaml (hunk 1 above)
Comment (matejvasek) [reply tensorflow/image-recognition/func.yaml]: I meant that this file shouldn't  be part of this PR at all.

**Comment by matejvasek** [comment]
I also think this should be under the python dir, right?

File: tensorflow/image-recognition/func.yaml (hunk 1 above)
Comment (matejvasek) [reply tensorflow/image-recognition/func.yaml]: This file should be deleted.

**Comment by matejvasek** [comment]
@chzhyang  Not related to this PR:
I don't know ML/Tensorflow. I wonder what is the character of the model binary with regard to change.
Does update to the model usually changes most of the content, or does it only change partially?
I am asking because I am working on `rsync` source upload for on-cluster-build, which is efficient for partial updates.

File: tensorflow/image-recognition/func.yaml (hunk 1 above)
Comment (chzhyang) [reply tensorflow/image-recognition/func.yaml]: Sorry, I will remove this file :)

**Comment by chzhyang** [comment]
> I also think this should be under the python dir, right?

OK, it will be moved to python dir

**Comment by chzhyang** [comment]
> @chzhyang Not related to this PR: I don't know ML/Tensorflow. I wonder what is the character of the model binary with regard to change. Does update to the model usually changes most of the content, or does it only change partially? I am asking because I am working on `rsync` source upload for on-cluster-build, which is efficient for partial updates.

Roughly speaking, the degree of model change varies [case by case](https://machinelearningmastery.com/update-neural-network-models-with-more-data/), related to changes of model architecture, data feature...but minmizing model change of model updating is a consistent pursuit(reduce cost) in deep learning. 
A model has many layers. Each layer represents different features and it can be imagined as a matrix composed of weights.
Most `fine-tuning` on base model just need to change serveral layers to update the representations of higher-order/lower-order feature, which means `partial` updates.  

fine-tuning:
- http://d2l.ai/chapter_computer-vision/fine-tuning.html
- https://www.tensorflow.org/tutorials/images/transfer_learning

Hope this can help you :)

**Review by lance — COMMENTED** [review]
@chzhyang thank you for this contribution! In order to work with `func create` the templates should be one directory below the target language/runtime. E.g. `python/tensorflow-image-recognition`. Unfortunately, the CLI cannot handle sub-directories such as `python/tensorflow/image-recognition` at the moment. Can you please move the template to a single directory under `python`?

File: python/tensorflow/image-recognition/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+This function project includes three test functions in [unit test](./test_func.py):
+- test GET request
+- test POST request
+- test empty request
+
+```bash
+$ python test_func.py
+```
+
+## Create funtion
```
[the remark this answers is not in the document]
Comment (chzhyang) [reply python/tensorflow/image-recognition/README.md:29]: Thanks, I will fix it and move the directory :)

File: python/tensorflow/image-recognition/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+Requirements:
+- [func](https://github.com/knative/func) is installed
+
+```bash
+$ cd func-tastic/python
+$ func create -l python tensorflow/image-recognition
+$ cd tensorflow/image_recognition
```
[the remark this answers is not in the document]
Comment (chzhyang) [reply python/tensorflow/image-recognition/README.md:36]: Thank you, docs will be updated

File: python/tensorflow/image-recognition/README.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+$ tree
+.
+└── image-recognition
+    ├── func.py
+    ├── func_test.py
+    ├── func.yaml
+    ├── Procfile
+    ├── README.md
+    └── requirements.txt
```
[the remark this answers is not in the document]
Comment (chzhyang) [reply python/tensorflow/image-recognition/README.md:45]: Thank you, addressed.

File: python/tensorflow/image-recognition/README.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+```
+Then add the logic of request handler and image recgnition inference to:
+- `func.py`
+- `image_recognition_service.py`
+
+## Build and run locally
+
+```bash
+$ func build -i docker.io/chzhyang/tensorflow-image-recognition:latest
```
[the remark this answers is not in the document]
Comment (chzhyang) [reply python/tensorflow/image-recognition/README.md:54]: addressed.

**Review by lance — CHANGES_REQUESTED** [review]
EDIT: Please disregard this. I realize that your function is not a cloudevent function and therefore does not need to have this annotation. However, something is still not working correctly. I'll need to dig further into this.


@chzhyang this is looking great! I tried it out locally, and realized that you'll need to have a `manifest.yaml` file in the root of `python/tensorflow-image-recognition` with the following contents in order for `func invoke` to work as expected.

```
invocation:
  format: cloudevent
```
Adding this `manifest.yaml` will override the defaults that are applied in a project's `func.yaml` when it is initially created. The default, if unspecified, is `http`.

**Review by lance — APPROVED** [review]
My issues seemed to be local. /lgtm

**Comment by matejvasek** [comment]
> > @chzhyang Not related to this PR: I don't know ML/Tensorflow. I wonder what is the character of the model binary with regard to change. Does update to the model usually changes most of the content, or does it only change partially? I am asking because I am working on `rsync` source upload for on-cluster-build, which is efficient for partial updates.
> 
> Roughly speaking, the degree of model change varies [case by case](https://machinelearningmastery.com/update-neural-network-models-with-more-data/), related to changes of model architecture, data feature...but minmizing model change of model updating is a consistent pursuit(reduce cost) in deep learning. A model has many layers. Each layer represents different features and it can be imagined as a matrix composed of weights. Most `fine-tuning` on base model just need to change serveral layers to update the representations of higher-order/lower-order feature, which means `partial` updates.
> 
> fine-tuning:
> 
> * http://d2l.ai/chapter_computer-vision/fine-tuning.html
> * https://www.tensorflow.org/tutorials/images/transfer_learning
> 
> Hope this can help you :)

@chzhyang could you please point me to some git repos where model changed?

**Comment by chzhyang** [comment]
> 



> > > @chzhyang Not related to this PR: I don't know ML/Tensorflow. I wonder what is the character of the model binary with regard to change. Does update to the model usually changes most of the content, or does it only change partially? I am asking because I am working on `rsync` source upload for on-cluster-build, which is efficient for partial updates.
> > 
> > 
> > Roughly speaking, the degree of model change varies [case by case](https://machinelearningmastery.com/update-neural-network-models-with-more-data/), related to changes of model architecture, data feature...but minmizing model change of model updating is a consistent pursuit(reduce cost) in deep learning. A model has many layers. Each layer represents different features and it can be imagined as a matrix composed of weights. Most `fine-tuning` on base model just need to change serveral layers to update the representations of higher-order/lower-order feature, which means `partial` updates.
> > fine-tuning:
> > 
> > * http://d2l.ai/chapter_computer-vision/fine-tuning.html
> > * https://www.tensorflow.org/tutorials/images/transfer_learning
> > 
> > Hope this can help you :)
> 
> @chzhyang could you please point me to some git repos where model changed?

TensorFlow gived a sample for model fine-tuning: 
- https://www.tensorflow.org/tutorials/images/transfer_learning
- https://github.com/tensorflow/docs/blob/master/site/en/tutorials/images/transfer_learning.ipynb

**Pull request merged**